core/src/main/java/io/questdb/metrics/QueryTracingJob.java

```diff
@@ -0,0 +1,110 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.metrics;
+
+import io.questdb.cairo.CairoEngine;
+import io.questdb.cairo.TableToken;
+import io.questdb.cairo.TableWriter;
+import io.questdb.griffin.SqlException;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.mp.SynchronizedJob;
+import io.questdb.mp.WorkerPool;
+import io.questdb.std.ValueHolderList;
+import io.questdb.std.str.Utf8StringSink;
+
+public class QueryTracingJob extends SynchronizedJob {
+    public static final String COLUMN_EXECUTION_MICROS = "execution_micros";
+    public static final String COLUMN_QUERY_TEXT = "query_text";
+    public static final String COLUMN_TS = "ts";
+    public static final String TABLE_NAME = "query_trace";
+    private static final int BATCH_LIMIT = 1024;
+    private static final int INITIAL_CAPACITY = 128;
+    private static final Log LOG = LogFactory.getLog(QueryTracingJob.class.getName());
+    private final ValueHolderList<QueryTrace> buffer;
+    private final CairoEngine engine;
+    private final MemCappedQueryTraceQueue queue;
+    private final QueryTrace trace = new QueryTrace();
+    private final Utf8StringSink utf8sink = new Utf8StringSink();
+    private TableToken tableToken;
+
+    public QueryTracingJob(CairoEngine engine) {
+        this.queue = engine.getMessageBus().getQueryTraceQueue();
+        this.buffer = new ValueHolderList<>(MemCappedQueryTraceQueue.ITEM_FACTORY, INITIAL_CAPACITY);
+        this.engine = engine;
+    }
+
+    public static void assignToPool(WorkerPool pool, CairoEngine engine) {
+        QueryTracingJob job = new QueryTracingJob(engine);
+        for (int i = 0, n = pool.getWorkerCount(); i < n; i++) {
+            pool.assign(i, job);
+        }
+    }
+
+    private void init() throws SqlException {
+        String fullName = engine.getConfiguration().getSystemTableNamePrefix() + TABLE_NAME;
+        engine.execute("CREATE TABLE IF NOT EXISTS '" + fullName +
+                "' (" + COLUMN_TS + " TIMESTAMP, " + COLUMN_QUERY_TEXT + " VARCHAR, " + COLUMN_EXECUTION_MICROS + " LONG)" +
+                " TIMESTAMP(" + COLUMN_TS + ") PARTITION BY HOUR TTL 1 DAY BYPASS WAL");
+        tableToken = engine.verifyTableName(fullName);
+    }
+
+    @Override
+    protected boolean runSerially() {
+        buffer.clear();
+        for (int i = 0; i < BATCH_LIMIT && queue.tryDequeue(buffer.peekNextHolder()); i++) {
+            buffer.commitNextHolder();
+        }
+        if (buffer.size() <= 0) {
+            return false;
+        }
+        TableWriter tableWriter0;
+        try {
+            tableWriter0 = engine.getWriter(tableToken, "query_tracing");
```

Comment:
writer must be retrieved and held by the job instance on init(), we must not lookup pool on every job iteration. Not only this is expensive, it is also allows someone else lock the writer and cripple SQL logging

Reply:
This implies I must assign the tracing job to a single thread, right? I thought it would be a better idea to assign it to all threads, so if any thread is stuck, another one can perform the work.

Reply:
Resolved in c84adca63c3a40d87e3520ad3eba150d9d07d82b